homeassistant/components/template/template_entity.py

```diff
@@ -0,0 +1,179 @@
+"""TemplateEntity utility class."""
+
+import logging
+from typing import Any, Callable, Optional, Union
+
+import voluptuous as vol
+
+from homeassistant.core import EVENT_HOMEASSISTANT_START, callback
+from homeassistant.exceptions import TemplateError
+from homeassistant.helpers.config_validation import match_all
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import Event, async_track_template_result
+from homeassistant.helpers.template import Template
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class _TemplateAttribute:
+    """Attribute value linked to template result."""
+
+    def __init__(
+        self,
+        entity: Entity,
+        attribute: str,
+        template: Template,
+        validator: Callable[[Any], Any] = match_all,
+        on_update: Optional[Callable[[Any], None]] = None,
+    ):
+        """Template attribute."""
+        self._entity = entity
+        self._attribute = attribute
+        self.template = template
+        self.validator = validator
+        self.on_update = on_update
+        self.async_update = None
+        self.add_complete = False
+
+    @callback
+    def async_setup(self):
+        """Config update path for the attribute."""
+        if self.on_update:
+            return
+
+        if not hasattr(self._entity, self._attribute):
+            raise AttributeError(f"Attribute '{self._attribute}' does not exist.")
+
+        self.on_update = self._default_update
+
+    @callback
+    def _default_update(self, result):
+        attr_result = None if isinstance(result, TemplateError) else result
+        setattr(self._entity, self._attribute, attr_result)
+
+    @callback
+    def _write_update_if_added(self):
+        if self.add_complete:
+            self._entity.async_write_ha_state()
+
+    @callback
+    def _handle_result(
+        self,
+        event: Optional[Event],
+        template: Template,
+        last_result: Optional[str],
+        result: Union[str, TemplateError],
+    ) -> None:
+        if isinstance(result, TemplateError):
+            _LOGGER.error(
+                "TemplateError('%s') "
+                "while processing template '%s' "
+                "for attribute '%s' in entity '%s'",
+                result,
+                self.template,
+                self._attribute,
+                self._entity.entity_id,
+            )
+            self.on_update(result)
+            self._write_update_if_added()
+
+            return
+
+        if not self.validator:
```

Comment:
This won't happen because `validator` defaults to `match_all`. Since we have this guard, should we drop the default value?

Reply:
Will do in a followup after I fix all the conflicts

Reply:
done here https://github.com/home-assistant/core/pull/39083